torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -0,0 +1,295 @@
+    MPI_CHECK(MPI_Comm_rank(MPI_COMM_WORLD, &rank_));
+
+  } catch (...) {
+    MPI_CHECK(MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE));
+  }
+
+  if (mpiThreadSupport_ != MPI_THREAD_MULTIPLE && numProcessGroups_ >= 1) {
+    throw std::runtime_error("More than one process group created, "
+                             "this is not supported due to the used MPI "
+                             "implementation doesn't provide the full support "
+                             "of multi-threading");
+  }
+  // increase the total PG count
+  ++numProcessGroups_;
+}
+
+void ProcessGroupMPI::destroy() {
+  std::unique_lock<std::mutex> lock(pgMutex_);
+
+  while (!queue_.empty()) {
+    queueConsumeCV_.wait(lock);
+  }
+  // Queue is empty, signal stop
+  stop_ = true;
+
+  // Release lock to allow threads to terminate
+  queueProduceCV_.notify_all();
+
+  lock.unlock();
+
+  // Join the single worker thread
+  workerThread_.join();
+
+  // Decrease the number of PG created
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  --numProcessGroups_;
+}
+
+void ProcessGroupMPI::abort() {
+  destroy();
+  MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
+}
+
+void ProcessGroupMPI::finalize() {
+  MPI_CHECK(MPI_Finalize());
+}
+
+void ProcessGroupMPI::runLoop() {
+  std::unique_lock<std::mutex> lock(pgMutex_);
+
+  while (!stop_) {
+    if (queue_.empty()) {
+      queueProduceCV_.wait(lock);
+      continue;
+    }
+
+    auto workTuple = std::move(queue_.front());
+
+    queue_.pop_front();
+    queueConsumeCV_.notify_one();
+
+    auto& workEntry = std::get<0>(workTuple);
+    auto& work = std::get<1>(workTuple);
+
+    lock.unlock();
+
+    try {
+      workEntry->run(workEntry);
+      work->finish();
+    } catch (const std::exception& ex) {
+      work->finishWithException(ex);
+    }
+
+    lock.lock();
+  }
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::enqueue(
+    std::unique_ptr<WorkEntry> entry) {
+  auto work = std::make_shared<WorkMPI>();
+  std::unique_lock<std::mutex> lock(pgMutex_);
+  queue_.push_back(std::make_tuple(std::move(entry), work));
+  queueProduceCV_.notify_one();
+  return work;
+}
+
+void ProcessGroupMPI::checkSingleTensor(
+    const std::vector<at::Tensor>& tensors) {
+  if (tensors.size() != 1) {
+    throw std::runtime_error("MPI process group only supports a single "
+                             "tensor op");
+  }
+  if (!tensors[0].is_contiguous()) {
+    throw std::runtime_error("input tensor has to be contiguous");
+  }
+  if (tensors[0].is_cuda() && !cudaAwareMpiCheck()) {
+    throw std::runtime_error("CUDA tensor detected and the MPI used doesn't "
+                             "have CUDA-aware MPI support");
+  }
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::broadcast(
+    std::vector<at::Tensor>& tensors,
+    const BroadcastOptions& opts) {
+  checkSingleTensor(tensors);
+  std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+    [&opts](std::unique_ptr<WorkEntry>& entry) {
```

Comment:
I don't think you have any guarantee that `opts` will be alive at the time `runFunc` will be called. Capture them as a value